Fix, test, and refactor concat()

kastiglione opened the pull request:
This fixes a crash in a `concat()` in cases where a signals completes before following signals are added to the queue. The crash happens while attempting to dequeue from the empty queue. This arose when doing a binary `concat` with the first signal being `empty`, which completes synchronously before the second signal is handled on the outer signal.

EDIT: Turned into a bunch of refactoring. See https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1714#issuecomment-73197857 for description.

cc @andymatuschak

kastiglione commented:
Found and fixed a logic bug where the producer returned from `concat()` would complete even when producers are still in the queue.

kastiglione commented:
I'm going to fill out the test stubs too…

kastiglione commented:
Ready for review.

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -475,10 +475,10 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
-						nextSignalProducer = state.queuedSignalProducers[0]
-						state.queuedSignalProducers.removeAtIndex(0)
+					if state.queuedSignalProducers.isEmpty {
+						completeIfAllowed(state)
```
I know this wasn't from your changes, but can you refactor this operator so that all invocations of `completeIfAllowed` occur _outside_ of the `modify` block? Otherwise, we'll be holding the spinlock for way longer than intended, and could even deadlock.

kastiglione replied:
> could even deadlock.

Is that for cycles? Or other ways too?

jspahrsummers replied:
Criss-crossing signals as well.

jspahrsummers commented on ReactiveCocoaTests/Swift/SignalProducerSpec.swift:
```diff
@@ -158,16 +158,126 @@ class SignalProducerSpec: QuickSpec {
 
 		describe("concat") {
-			pending("should start subsequent inner signals upon completion") {
+			context("sequencing") {
```
Can you please use `describe` instead of `context`, for consistency?

kastiglione replied:
Just for this use of `context`, or the other use I had too?

jspahrsummers commented on ReactiveCocoaTests/Swift/SignalProducerSpec.swift:
```diff
@@ -158,16 +158,126 @@ class SignalProducerSpec: QuickSpec {
+
+				beforeEach {
+					outerProducer = SignalProducer { observer, _ in
+						outerSink = observer
```
Can this use `SignalProducer.buffer` or something instead?

kastiglione replied:
Aaaaaah… I wondered if there was something analogous to `pipe` for `SignalProducer`, yet I've skipped over `buffer` dozens of times on the name alone :see_no_evil:

jspahrsummers commented:
Made a few notes which might significantly change the code, so I'll pause here.

kastiglione commented:
:100:

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -443,18 +443,12 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
-			} else {
-				return false
-			}
+		let complete💯: Void -> Void = {
```
nopenopenope

jspahrsummers commented:
:angry:

kastiglione commented:
> :angry: 

Sorry, eh :disappointed: 

I saw an email with:

> Criss-crossing signals as well.

Was that related to an older revision?

kastiglione commented:
Did a big 'ol bunch of refactoring. Kinda like how it's turned out so far, but eager to hear feedback.

Table of Contents, with more info below.

1. Wanted to get rid of the recursive closure and the retain cycle uncertainty.
1. Wanted to make the atomic operations on `ConcatState` its own responsibility.
1. Got rid of the boolean pair `selfCompleted` and `latestSignalCompleted`.

Re 1: With the `Z` combinator, I wasn't entirely sure the `completeIfAllowed` closure needed to strongly capture the `subscribeToSignalProducer` block, nor that it needed to `nil` it out. I _think_ the recursion and curried nature of `Z()` took care of, but instead of proving this, I figured it would be simpler to make `subscribeToSignalProducer` a free function that had no free variables. Then it could refer to itself no problem. To simplify calling `subscribeToSignalProducer`, I added some immutable state to `ConcatState`, specifically the concat observer and disposable.

Re 2: Not much to say on this, check out locking inside `ConcatState`.

Re 3: I quite liked how this played out. In order to get rid of the `selfCompleted` flag, I made it so that upon completion of the outer producer, one more inner producer is added to the queue. This last producer is just `.empty` with an `on(:completed)` side effect of calling `sendCompleted(observer)`. Next, to get rid of the `latestSignalCompleted` flag, I changed the semantics of the queue such that the active producer stays in the queue until it is completed. Thus, if the queue is empty, then the last producer *has* completed, and if it is not empty, then the last producer *has not* completed.

The result is that this implementation moved logic from being mostly inline in the `SignalProducer` definitions, to being responsibility of `ConcatState`.

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -521,21 +458,71 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
+		lock()
+		// An empty queue means the concat is idle, ready & waiting to start the
+		// next producer given.
+		shouldStart = self.queuedSignalProducers.isEmpty
```
Can you please remove references to `self` that are unnecessary?

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -521,21 +458,71 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
+
+	var spinlock = OS_SPINLOCK_INIT
+	func lock() { withUnsafeMutablePointer(&spinlock, OSSpinLockLock) }
+	func unlock() { withUnsafeMutablePointer(&spinlock, OSSpinLockUnlock) }
```
Can we wrap the array in an `Atomic` instead of doing this?

kastiglione replied:
![](http://www.reactiongifs.com/wp-content/uploads/2012/07/haters_gonna_duh.gif)

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -521,21 +458,71 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
+		let serialDisposable = SerialDisposable()
+		let serialDisposableCompositeHandle = disposable.addDisposable(serialDisposable)
+
+		serialDisposable.innerDisposable = signalProducer.start(next: { value in
```
If you use `startWithSignal`, the serial disposable should be unnecessary.

jspahrsummers commented:
Very nice! Just a few more notes.

kastiglione commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -521,21 +458,68 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
+	/// Subscribes to the given signal producer.
+	func startNextSignalProducer(signalProducer: SignalProducer<T, E>) {
+		signalProducer.startWithSignal { signal, disposable in
+			self.disposable.addDisposable(disposable)
```
@jspahrsummers Does this do it?

jspahrsummers replied:
Yep!

kastiglione commented:
:ferris_wheel:

jspahrsummers commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -521,21 +458,68 @@ public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> S
+	let disposable: CompositeDisposable
+
+	/// The active producer, if any, and the producers waiting to be started.
+	var queuedSignalProducers: Atomic<[SignalProducer<T, E>]> = Atomic([])
```
Can you make this `let`, since it's an immutable reference to mutable data?

jspahrsummers commented:
Just one more note, and then feel free to merge this yourself. :roller_coaster:
